src/MagnumPlugins/KtxImageConverter/KtxImageConverter.cpp

```diff
@@ -0,0 +1,932 @@
+    using namespace Containers::Literals;
+
+    const Containers::Pair<Containers::StringView, Containers::StringView> keyValueMap[]{
+        /* Origin left, bottom, back (increasing right, up, out) */
+        /** @todo KTX spec says "most other APIs and the majority of texture
+                  compression tools use [r, rd, rdi]".
+                  Should we just always flip image data? Maybe make this
+                  configurable. */
+        Containers::pair("KTXorientation"_s, "ruo"_s.prefix(dimensions)),
+        Containers::pair("KTXwriter"_s, "Magnum::KtxImageConverter 1.0"_s)
+    };
+
+    /* Calculate size */
+    std::size_t kvdSize = 0;
+    for(const auto& entry: keyValueMap) {
+        const UnsignedInt length = entry.first().size() + 1 + entry.second().size() + 1;
+        kvdSize += sizeof(length) + (length + 3)/4*4;
+    }
+    CORRADE_INTERNAL_ASSERT(kvdSize % 4 == 0);
+
+    /* Pack. We assume that values are text strings, no endian-swapping needed. */
+    std::size_t kvdOffset = 0;
+    Containers::Array<char> keyValueData{ValueInit, kvdSize};
+    for(const auto& entry: keyValueMap) {
+        const auto key = entry.first();
+        const auto value = entry.second();
+        const UnsignedInt length = key.size() + 1 + value.size() + 1;
+        *reinterpret_cast<UnsignedInt*>(keyValueData.suffix(kvdOffset).data()) = length;
+        Utility::Endianness::littleEndianInPlace(length);
+        kvdOffset += sizeof(length);
+        Utility::copy(key, keyValueData.suffix(kvdOffset).prefix(key.size()));
+        kvdOffset += entry.first().size() + 1;
+        Utility::copy(value, keyValueData.suffix(kvdOffset).prefix(value.size()));
+        kvdOffset += entry.second().size() + 1;
+        kvdOffset = (kvdOffset + 3)/4*4;
+    }
+    CORRADE_INTERNAL_ASSERT(kvdOffset == kvdSize);
+
+    /* Fill level index */
+    const Math::Vector<dimensions, Int> size = imageLevels.front().size();
+    if(size.product() == 0) {
+        Error() << "Trade::KtxImageConverter::convertToData(): image for level 0 is empty";
+        return {};
+    }
+
+    const UnsignedInt numMipmaps = Math::min<UnsignedInt>(imageLevels.size(), Math::log2(size.max()) + 1);
+    if(imageLevels.size() > numMipmaps)
+        Warning{} << "Trade::KtxImageConverter::convertToData(): expected at most" <<
+            numMipmaps << "mip level images but got" << imageLevels.size() << Debug::nospace <<
+            ", extra images will be ignored";
+
+    Containers::Array<Implementation::KtxLevel> levelIndex{numMipmaps};
+
+    const std::size_t levelIndexSize = numMipmaps*sizeof(Implementation::KtxLevel);
+    std::size_t levelOffset = sizeof(Implementation::KtxHeader) + levelIndexSize +
+        dataFormatDescriptor.size() + keyValueData.size();
+
+    /** @todo Handle block-compressed formats. Need block size and block count
+              in each dimension. How should skip be handled, if it's not a
+              multiple of the block size? */
```

Comment:
Same as I said above, can you for now just put an (internal) assert for skip being zero and row length / image height not defined? When I add what's needed to the CompressedImage APIs, I'll update the code here.